classes/battle/Battle.php

```diff
@@ -294,33 +305,77 @@ public function isPreparationPhase(): bool {
      * @throws RuntimeException
      */
 
-    public function timeRemaining(): int {
-        return Battle::TURN_LENGTH - (time() - $this->turn_time);
+    public function timeRemaining($player_id = 0): int {
+        // If not set, use current player
+        if ($player_id == 0) {
+            $player_id = $this->player->id;
+        }
+        // If action is set, player's turn time is frozen - else modify by time since last turn
+        switch ($player_id) {
+            case $this->player1_id:
+                if (isset($this->fighter_actions[$this->player1->combat_id])) {
+                    return $this->player1_time;
+                }
+                return $this->player1_time - (time() - $this->turn_time);
+            case $this->player2_id:
+                if (isset($this->fighter_actions[$this->player2->combat_id])) {
+                    return $this->player2_time;
+                }
+                return $this->player2_time - (time() - $this->turn_time);
+            default:
+                return 0;
+        }
     }
 
     public function prepTimeRemaining(): int {
         return Battle::PREP_LENGTH - (time() - $this->start_time);
     }
 
+    public function updatePlayerTime($player_id = 0, $min = false) {
+        // if min, set to 15 seconds - else add 15 seconds and bound to min/max turn time
+        switch ($player_id) {
+            case $this->player1_id:
+                if ($min) {
+                    $this->player1_time = self::TURN_LENGTH;
+                } else {
+                    $this->player1_time = min(max(($this->player1_time - (time() - $this->turn_time)) + self::TURN_LENGTH, self::TURN_LENGTH), self::INITIAL_TURN_LENGTH);
```

Comment:
Can we make a quick System helper function for this? like

```php
function clampNumber($number, $min, $max) {
   return max(min($number, $max), $min);
}
```
```suggestion
                    $this->player1_time = System::clampNumber(
                       number: $this->player1_time - (time() - $this->turn_time)) + self::TURN_LENGTH, 
                       min: self::TURN_LENGTH, 
                       max: self::INITIAL_TURN_LENGTH
                    );
```